### ISSUE TITLE
Salt and Options managment

### ISSUE THREAD

**Issue by Darlio** [kosprov/jargon2-api#2]
Hello and first of all congratulations !

This in not a issue... more about understanding correctly your implementation.

I'm going to make some statements and would like you to just confirm or not what I understood. Probably what I'm about to write could be a great boost to your implementation popularity, so you could put (in your words) in the documentations.

Please tell me if what is next is correct ?

--------------------------

1. Why the SALT is not part of the parameters that should be given during hashing / verification ?

Because the SALT is automatically generated the first time, and integrated into the final result of hashing. Rather it being added at beginning / end / or even blended in the password prior hashing the resulting chars[] is relieved from the DEVELOPER which can concentrate on more important things. 

2. If I change the parameters of the hashing process, once I have already a certain amount of saved salted and hashed password in my DataBase, will I still be able to verify them?

Yes. As you can notice in the simple example, instantiation of the hasher is done specifying a certain number of parameters:
![image](https://user-images.githubusercontent.com/32061339/43312188-d49a6dfc-918c-11e8-8215-b5595f8a535d.png)

These parameters are also INCLUDED (as is the SALT) inside the resulting hash.
So the verification can be done using these specific parameters (extraction of the correct SALT length for example), no matter how different the actual hash object was instantiated.

As an addition to the above, this clever way of doing a "SAVE" of parameters for each hash, also gives the possibility to migrate smoothly the passwords to the new parameters during authentication of users (see the 
![image](https://user-images.githubusercontent.com/32061339/43312281-0f93cf16-918d-11e8-8467-7aa1c5ea0f4a.png)
 example and followed explanation)

--------------------------

**Comment by kosprov** [comment]
Thanks for you time in evaluating this library.

**TL;DR** : You are right on both your statements.

These were stuff I tried to understand myself when I first experimented with Argon2. One of the reasons I wrote another Argon2 wrapper API was that I was getting confused by other libraries, even the RI itself (e.g. [why do I need to pass the type when verifying encoded hashes](https://github.com/P-H-C/phc-winner-argon2/issues/234))

I'll try to explain as best as possible and give full answers to your questions. I know you've already figured out most of the things that follow. I'll still write them for reference.

**Background:**

Argon2 specifies an _encoded hash_ format which contains everything needed for verification. The syntax is:

    $argon2<type>[$v=<version>]$m=<memoryCost>,t=<timeCost>,p=<lanes>$<salt>$<hash>

for example:

    $argon2i$v=19$m=65536,t=2,p=4$c29tZXNhbHQ$RdescudvJCsgt3ub+b+dWRWJTmaaJObG

This string tells:

    Type: Argon2i
    Version: 1.3
    Memory cost: 64MB
    Time cost: 2
    Parallelism: 4
    Salt: c29tZXNhbHQ (in Base64 without padding)
    Hash: RdescudvJCsgt3ub+b+dWRWJTmaaJObG (in Base64 without padding)

From the length of the Base64 value, we can conclude that the hash length was 24 bytes. This length is also need in verification but it's not encoded separately. You can just count the bytes.

Given this string alone, we can retrieve all options, the salt and the output length and re-calculate a hash. Then, we compare byte-by-byte with `RdescudvJCsgt3ub+b+dWRWJTmaaJObG` to see if all bytes match.

Encoded hash format has nothing to do with Jargon2. Other Argon2 implementations would be able to verify encoded hashes produced by Jargon2 and vice-versa.

So, when you store encoded hashes, you don't need any other information to do the verification. The only exception is when you're doing keyed-hashing where you have to get access to the same key. This key is meant to be kept secret, so you never include it in the encoded hash or store it in the same database etc.

As far as the salt value is concerned, it must be random and unique __per hash__ for two reasons:

1. If 2 users happen to choose the same password, you would not be able to tell just by looking at the hashes. With random and unique salts per hash, identical passwords have different hashes.

2. If an attacker gets access to your password database and attempts a dictionary attack, he would have to compute different rainbow tables for each password. That fact, combined with the "memory hard" property inherent in Argon2, makes such an attack very slow.

**Now, your questions:**

> 1. Why the SALT is not part of the parameters that should be given during hashing / verification ?

During hashing, the salt is automatically generated internally, as this is what's needed in most of the use cases: a random and unique salt per hash. You can pass a salt you've generated by yourself (see example below), but this is a rather rare need.

During verification, the salt is found in the encoded hash, as well as all other Argon2 options.

Since we are on the salt subject, I want to spend some time on what are the options on dealing with salts and salt generation in Jargon2. 

In the majority of cases, the salt needs to be randomly generated for every hash (the rationale above). 

When you write:

    Hasher hasher = jargon2Hasher()
                .type(type)
                .memoryCost(memoryCost)
                .timeCost(timeCost)
                .parallelism(parallelism)
                .hashLength(hashLength);

    String encodedHash1 = hasher.password(password).encodedHash();
    String encodedHash2 = hasher.password(password).encodedHash();
    String encodedHash3 = hasher.password(password).encodedHash();

all encoded hashes have different salts and different output. That happens because you start from the prototype `hasher` which does not encapsulate any salt. When `encodedHash()` gets executed, it generates a new salt and produces the result. Then, everything gets discarded and, on the next line, you start over from the prototype `hasher` which does not encapsulate any salt. You'll see more on this later.

Salt manipulation in a `Hasher` can be configured in two ways:

1. Set a different salt generator. The default one is a singleton `java.security.SecureRandom` instantiated with your JVMs settings. By calling any of the `hasher.saltGenerator()` methods, you can change the PRNG algorithm, the provider or even give your own implementation of `com.kosprov.jargon2.api.Jargon2.SaltGenerator`.

2. You can provide the salt "as-is" and set it to the hasher. Then, it will use this value and won't generate one internally. Let's say you want to generate a salt but without implementing a salt generator. 

The code above can be re-written as:

    Hasher hasher = jargon2Hasher()
                .type(type)
                .memoryCost(memoryCost)
                .timeCost(timeCost)
                .parallelism(parallelism)
                .hashLength(hashLength);

    String encodedHash1 = hasher.salt(generateSalt()).password(password).encodedHash();
    String encodedHash2 = hasher.salt(generateSalt()).password(password).encodedHash();
    String encodedHash3 = hasher.salt(generateSalt()).password(password).encodedHash();

Again, all hashes will have different salts and different output. This is because you don't change the prototype `hasher` (if you recall, they are immutable). You set the salt to an intermediary `hasher`, you set the password to another intermediary `hasher` and you ask for the encoded hash. At that point, the salt is already there and is not generated internally. On the next line, you start again over from the prototype `hasher` so everything that happened so far is garbage collectible.

BUT, if you do this:

    Hasher hasher = jargon2Hasher()
                .type(type)
                .memoryCost(memoryCost)
                .timeCost(timeCost)
                .parallelism(parallelism)
                .hashLength(hashLength)
                .salt(generateSalt()); // Generate and set a salt on the prototype hasher

    String encodedHash1 = hasher.password(password).encodedHash();
    String encodedHash2 = hasher.password(password).encodedHash();
    String encodedHash3 = hasher.password(password).encodedHash();

... all encoded hashes would be the same. There is no specific reason on why one would want to do this ~~, unless he's writting a dictionary attack program~~ .

Bottom line: You want random and unique salt per hash and you can achieve this if you don't explicitly set a salt on the prototype `hasher`.

> 2. If I change the parameters of the hashing process, once I have already a certain amount of saved salted and hashed password in my DataBase, will I still be able to verify them?

Yes, for encoded hashes (as opposed to raw hashes). 

Raw hashes are just the output bytes (the `RdescudvJCsgt3ub+b+dWRWJTmaaJObG` in the example) and you need to manage options and salt storage by yourself (e.g. on separate database columns).

And yes, as mentioned in the javadoc of `Hasher#propertiesMatch(String)`:

> Tests whether this hasher configuration matches with properties found encoded in the given hash.

This will tell you if an old hash needs migration. Obviously, you test it on the prototype `hasher` instance after you've loaded all your new options.
